tiledb/sm/array_schema/test/unit_attribute.cc

```diff
@@ -44,164 +44,308 @@ TEST_CASE("ANY Attribute default cell val num", "[array_schema][attribute]") {
+      Datatype::FLOAT64,
+      Datatype::INT8,
+      Datatype::UINT8,
+      Datatype::INT16,
+      Datatype::UINT16,
+      Datatype::INT32,
+      Datatype::UINT32,
+      Datatype::INT64,
+      Datatype::UINT64,
       Datatype::STRING_ASCII,
-      constants::var_num,
-      DataOrder::INCREASING_DATA);
-  CHECK_THROWS(a.set_cell_val_num(1));
-  CHECK_THROWS(a.set_cell_val_num(100));
-  CHECK(a.cell_val_num() == constants::var_num);
+      Datatype::STRING_UTF8,
+      Datatype::STRING_UTF16,
+      Datatype::STRING_UTF32,
+      Datatype::STRING_UCS2,
+      Datatype::STRING_UCS4,
+      Datatype::DATETIME_YEAR,
+      Datatype::DATETIME_WEEK,
+      Datatype::DATETIME_DAY,
+      Datatype::DATETIME_HR,
+      Datatype::DATETIME_MIN,
+      Datatype::DATETIME_SEC,
+      Datatype::DATETIME_MS,
+      Datatype::DATETIME_US,
+      Datatype::DATETIME_NS,
+      Datatype::DATETIME_PS,
+      Datatype::DATETIME_FS,
+      Datatype::DATETIME_AS,
+      Datatype::TIME_HR,
+      Datatype::TIME_MIN,
+      Datatype::TIME_SEC,
+      Datatype::TIME_MS,
+      Datatype::TIME_US,
+      Datatype::TIME_NS,
+      Datatype::TIME_PS,
+      Datatype::TIME_FS,
+      Datatype::TIME_AS,
+      Datatype::BOOL,
+      Datatype::BLOB,
+      Datatype::GEOM_WKT,
+      Datatype::GEOM_WKB);
+
+  const unsigned int valid_cell_val_num =
+      GENERATE(1, 2, 5, 10, 100, 1000, constants::var_num);
+
+  DYNAMIC_SECTION(
+      "Constructor: default cell val num is 1 for datatype " +
+      datatype_str(attribute_datatype)) {
+    CHECK(Attribute("a", attribute_datatype, false).cell_val_num() == 1);
+  }
+
+  DYNAMIC_SECTION(
+      "Constructor: all nonzero cell val nums are valid for unordered "
+      "attribute with "
+      "datatype " +
+      datatype_str(attribute_datatype)) {
+    CHECK(
+        Attribute(
+            "a",
+            attribute_datatype,
+            valid_cell_val_num,
+            DataOrder::UNORDERED_DATA)
+            .cell_val_num() == valid_cell_val_num);
+  }
+
+  DYNAMIC_SECTION(
+      "Constructor: zero cell val num is invalid for datatype " +
+      datatype_str(attribute_datatype)) {
+    CHECK_THROWS(
+        Attribute("a", attribute_datatype, 0, DataOrder::UNORDERED_DATA));
+  }
+
+  DYNAMIC_SECTION(
+      "set_cell_val_num: all nonzero cell val nums are valid for unordered "
+      "attribute "
+      "with datatype " +
+      datatype_str(attribute_datatype)) {
+    Attribute a("a", attribute_datatype, 1, DataOrder::UNORDERED_DATA);
+    REQUIRE(a.cell_val_num() == 1);
+    REQUIRE_NOTHROW(a.set_cell_val_num(valid_cell_val_num));
+    CHECK(a.cell_val_num() == valid_cell_val_num);
+  }
+
+  DYNAMIC_SECTION(
+      "set_cell_val_num: zero cell val num is invalid for datatype " +
+      datatype_str(attribute_datatype)) {
+    Attribute a(
+        "a", attribute_datatype, valid_cell_val_num, DataOrder::UNORDERED_DATA);
+    REQUIRE(a.cell_val_num() == valid_cell_val_num);
+    REQUIRE_THROWS(a.set_cell_val_num(0));
```

Comment:
>  the generator that governs this doesn't cover ANY

This is intentional as all the ANY tests are covered by the separate test cases.

> note that this generator is copypasta below

Unfortunately it actually is not, the second datatype generation does not include string types since they (well, `STRING_ASCII` at least) have some different semantics enumerated in a different `TEST_CASE`.

(that is something to improve here, is make sure that the non-ascii string types are properly covered by one of the generators)

Reply:
> that is something to improve here, is make sure that the non-ascii string types are properly covered by one of the generators

I gave this a shot, they all return something along the lines of `Datatype 'STRING_UTF32' is not a valid datatype for an ordered attribute`.  I don't think they are right to include in this test case but I'll leave a comment.